smart-contract-sale/src/contract.rs

```diff
@@ -0,0 +1,1017 @@
+use cosmwasm_std::{
+    attr, entry_point, to_binary, BankMsg, Binary, Coin, Deps, DepsMut, Env, MessageInfo, Response,
+    StdResult, WasmMsg,
+};
+use cw2::get_contract_version;
+use provwasm_std::{Marker, MarkerType, ProvenanceMsg, ProvenanceQuerier, ProvenanceQuery};
+
+use crate::error::{contract_err, ContractError};
+use crate::msg::{DCCWasmMsg, ExecuteMsg, QueryMsg, Validate};
+use crate::state::{config_read, get_sale_storage, get_sale_storage_read, Sale, Status};
+
+pub const CRATE_NAME: &str = env!("CARGO_CRATE_NAME");
+pub const PACKAGE_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// smart contract execute entrypoint
+#[entry_point]
+pub fn execute(
+    deps: DepsMut<ProvenanceQuery>,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<ProvenanceMsg>, ContractError> {
+    msg.validate()?;
+
+    match msg {
+        ExecuteMsg::CreateSale { id, buyer, price } => create_sale(deps, info, id, buyer, price),
+        ExecuteMsg::CompleteSale { id } => complete_sale(deps, info, id),
+        ExecuteMsg::CancelSale { id } => cancel_sale(deps, info, id),
+    }
+}
+
+fn create_sale(
+    deps: DepsMut<ProvenanceQuery>,
+    info: MessageInfo,
+    id: String,
+    buyer: String,
+    price: Coin,
+) -> Result<Response<ProvenanceMsg>, ContractError> {
+    // validate params
+    if info.funds.is_empty() {
```

Comment:
so if we were selling a loan that would be the funds passed in correct?

Reply:
yes there should be one coin as funds